core/state_processor.go

```diff
@@ -129,9 +129,16 @@ func applyTransaction(msg types.Message, config *params.ChainConfig, gp *GasPool
 	receipt.TxHash = tx.Hash()
 	receipt.GasUsed = result.UsedGas
 
+	nonce := tx.Nonce()
+	if msg.IsDepositTx() && config.IsOptimismRegolith(evm.Context.Time) {
+		// Subtract 1 as the nonce in state has already been incremented from what was used
+		nonce = statedb.GetNonce(msg.From()) - 1
```

Comment:
By calling the state-db here, and subtracting one, we assume the account is an EOA. If it was a contract, then it could create multiple other contracts, and increment its nonce by more than 1.

Theoretically it's safe to assume it's an EAO because on L1 we add a modifier to the depositor from address to make it unique from a contract address, but just in case we change this behavior, I would prefer to do the statedb nonce load before the EVM processing to not have to do any `-1` on it, and thus make sure it's the correct value regardless of EOA/contract behavior. With future account-abstraction changes this may also be worthwhile.

Reply:
Yep, good catch. It's easy to grab the value prior to transition and avoids the need to explain why we're subtracting 1.